### PR TITLE
Sliding Window + Long-Context Training: val_bpb=1.1764

### PR DESCRIPTION
**Pull request by saml212**
## Score

**val_bpb = 1.1764** (baseline: 1.2244, improvement: 0.048 BPB / 0.087 nats)

Across seeds: 1.1764 (SEED=1337)

## Approach

Train at longer sequences (2048 tokens) with high Muon momentum (0.99), low learning rate (0.02), and tight gradient clipping (0.3). Evaluate with overlapping sliding windows where every scored token sees 1536+ tokens of preceding context.

### 1. Training length doesn't matter with sliding window eval
Training at 2048 vs 4096 gives identical BPB when evaluated with sliding window (1.1764 vs 1.1765). The sliding window already provides long context at eval — the model just needs to learn local patterns well. Training at 2048 is strictly better because it gets more optimization steps in 10 minutes.

### 2. Gradient clipping sweet spot for long sequences
Long-sequence training benefits from a narrow clipping window (0.3 vs default 1.0). Full sweep from 0.0 to 1.0 identified 0.3 as optimal — stabilizes long-sequence gradient variance without over-constraining.

### 3. Batch=786K is optimal for train@2048
Swept 393K to 1M. The sweet spot (786K) balances gradient noise against step count within the 10-min budget.

### 4. Quantization-aware warmdown (from our earlier PR #61)
Aggressive LR warmdown reduces post-quant penalty 3x (0.014→0.005 BPB). However, this interacts with base LR — the benefit only appears at high LR (0.06), not low LR (0.02). Full curve mapped across 10 warmdown values in PR #61.

## Configuration

```
TRAIN_SEQ_LEN=2048 TRAIN_BATCH_TOKENS=786432 MATRIX_LR=0.02 SCALAR_LR=0.02
TIED_EMBED_LR=0.03 MUON_MOMENTUM=0.99 MUON_MOMENTUM_WARMUP_START=0.92
MUON_MOMENTUM_WARMUP_STEPS=1500 WARMDOWN_ITERS=3000 GRAD_CLIP_NORM=0.3
EVAL_SEQ_LEN=2048 EVAL_STRIDE=512
```

## Reproduction

```bash
TRAIN_SEQ_LEN=2048 TRAIN_BATCH_TOKENS=786432 MATRIX_LR=0.02 SCALAR_LR=0.02 \
TIED_EMBED_LR=0.03 MUON_MOMENTUM=0.99 MUON_MOMENTUM_WARMUP_START=0.92 \
MUON_MOMENTUM_WARMUP_STEPS=1500 WARMDOWN_ITERS=3000 GRAD_CLIP_NORM=0.3 \
EVAL_SEQ_LEN=2048 EVAL_STRIDE=512 \
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

15.88MB artifact. 8xH100 SXM (RunPod, ~47ms/step).

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Sliding Window + Long-Context Training: val_bpb=1.1764

**BPB:** 1.1764 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `50225c568f63`, file `records/track_10min_16mb/2026-03-19_WarmdownQuantization/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.02s, dim=512, layers=9, vocab=1024, code=52510 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.02s, dim=512, layers=9, vocab=1024, code=52510 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
